# Python für Verfahrenstechniker — Teil 1
## Erste Schritte: Zahlen, Variablen, Datentypen

Diese Notebook-Reihe führt in vier Teilen in das Programmieren mit Python ein. Alle Beispiele
stammen aus dem Verfahrenstechnik-Alltag — Sie lernen Python also nicht abstrakt, sondern an
Aufgaben, die Sie sonst mit Taschenrechner und Formelsammlung lösen.

**Die vier Teile:**

| Teil | Inhalt |
|------|--------|
| **1** | Erste Schritte: Zahlen, Variablen, Datentypen (dieses Notebook) |
| 2 | Kontrollstrukturen, Funktionen und Fehlersuche |
| 3 | NumPy und grafische Darstellung |
| 4 | Messdaten, gute Praxis und Abschlussprojekt Batch-Reaktor |

---

### Lernziele dieses Notebooks

Nach diesem Notebook können Sie

- ein Jupyter Notebook bedienen und den Unterschied zwischen Code- und Markdown-Zellen erklären,
- Python als Taschenrechner benutzen und Zwischenergebnisse in Variablen speichern,
- die Datentypen `int`, `float`, `bool` und `str` unterscheiden,
- Ergebnisse mit `f`-Strings sauber und mit Einheiten formatiert ausgeben,
- Listen und Dictionaries erzeugen und darauf zugreifen.

**Zeitaufwand:** ca. 90 Minuten inklusive Übungen.

---
## 1. Wie ein Notebook funktioniert

Ein Jupyter Notebook besteht aus **Zellen**. Es gibt zwei Sorten, die Sie unterscheiden müssen:

- **Code-Zellen** enthalten Python-Anweisungen. Sie werden ausgeführt und liefern ein Ergebnis.
- **Markdown-Zellen** enthalten Text, Formeln und Tabellen — wie diese hier. Sie werden nur gesetzt.

Eine Zelle führen Sie mit **`Shift` + `Enter`** aus. Probieren Sie es an der nächsten Zelle aus.

In [1]:
2 + 3

5

Unter der Zelle erscheint das Ergebnis. Jupyter zeigt automatisch den Wert des **letzten Ausdrucks**
einer Zelle an. Wollen Sie mehrere Werte sehen, brauchen Sie `print()`:

In [2]:
print(2 + 3)
print(2 * 3)
print(2 ** 3)   # Potenz, nicht ^ !

5
6
8


### Die wichtigste Warnung vorweg: der Kernel

Hinter dem Notebook läuft ein **Kernel** — ein Python-Prozess, der sich alles merkt, was Sie
ausgeführt haben. Entscheidend ist die **Reihenfolge der Ausführung**, nicht die Reihenfolge auf
dem Bildschirm. Wenn Sie eine Zelle weiter oben nachträglich ändern, wirkt sich das nicht rückwirkend
aus.

Die Zahl in `[ ]` links neben einer Code-Zelle zeigt, als wievielte sie ausgeführt wurde.

> **Faustregel:** Wenn sich Ihr Notebook seltsam verhält, wählen Sie
> *Kernel → Restart Kernel and Run All Cells*. Damit läuft alles einmal sauber von oben durch.
> Bevor Sie ein Notebook abgeben, sollten Sie das immer tun.

---
## 2. Python als Taschenrechner

Die Grundrechenarten funktionieren so, wie Sie es erwarten. Zwei Operatoren sind neu:

| Operator | Bedeutung | Beispiel | Ergebnis |
|----------|-----------|----------|----------|
| `+ - * /` | Grundrechenarten | `7 / 2` | `3.5` |
| `**` | Potenz | `2 ** 10` | `1024` |
| `//` | Ganzzahldivision (abgerundet) | `7 // 2` | `3` |
| `%` | Rest (Modulo) | `7 % 2` | `1` |

Klammern gelten wie in der Mathematik, Punkt vor Strich ebenfalls.

In [3]:
print(7 / 2)     # normale Division -> Kommazahl
print(7 // 2)    # Ganzzahldivision -> abgerundet
print(7 % 2)     # Rest
print(2 ** 0.5)  # Wurzel als Potenz mit 0.5

3.5
3
1
1.4142135623730951


Für Wurzeln, Logarithmen und Winkelfunktionen brauchen Sie das Modul `math`. Ein Modul ist eine
Sammlung fertiger Funktionen, die Sie erst **importieren** müssen:

In [4]:
import math

print(math.sqrt(2))       # Quadratwurzel
print(math.log(100))      # natürlicher Logarithmus (ln)
print(math.log10(100))    # Zehnerlogarithmus
print(math.exp(1))        # e-Funktion
print(math.pi)            # Kreiszahl

1.4142135623730951
4.605170185988092
2.0
2.718281828459045
3.141592653589793


> **Achtung:** `math.log()` ist der **natürliche** Logarithmus, nicht der Zehnerlogarithmus.
> Das ist eine klassische Fehlerquelle beim Umrechnen von Antoine-Konstanten.

---
## 3. Variablen

Ein Ergebnis, das Sie weiterverwenden wollen, speichern Sie in einer **Variablen**. Das Gleichheitszeichen
ist dabei keine Gleichung, sondern eine **Zuweisung**: Der Wert rechts wird berechnet und links gespeichert.

In [5]:
# Zustand eines Gases
p = 2.5e5      # Druck in Pa   (2.5e5 bedeutet 2.5 * 10^5)
T = 350.0      # Temperatur in K
V = 0.040      # Volumen in m^3
R = 8.314      # allgemeine Gaskonstante in J/(mol K)

# ideales Gasgesetz nach der Stoffmenge aufgelöst: n = p*V / (R*T)
n = p * V / (R * T)

n

3.436544211141276

### Regeln für Variablennamen

- erlaubt sind Buchstaben, Ziffern und der Unterstrich `_`, aber **keine Ziffer am Anfang**
- Groß- und Kleinschreibung wird unterschieden: `T` und `t` sind zwei verschiedene Variablen
- keine Umlaute, keine Leerzeichen, keine Sonderzeichen
- in Python schreibt man Variablen klein und trennt Wörter mit `_`: `massenstrom_ein`

Gute Namen sind in der Verfahrenstechnik besonders wichtig, weil Sie mit vielen ähnlichen Größen
arbeiten. `T_ein` und `T_aus` versteht man auch nach drei Monaten noch, `a` und `b` nicht.

In [6]:
# Nachvollziehbar:
T_ein = 293.15      # K
T_aus = 348.15      # K
delta_T = T_aus - T_ein

print(delta_T)

55.0


> **Wichtig:** Python kennt **keine Einheiten**. Die Zahl 350 ist für Python einfach 350 — ob das
> Kelvin oder Grad Celsius sind, weiß nur Sie. Notieren Sie die Einheit deshalb *immer* als Kommentar
> hinter der Zuweisung. Ein guter Teil aller Fehler in technischen Rechnungen sind Einheitenfehler.

---
## 4. Datentypen

Jeder Wert in Python hat einen **Typ**. Für den Anfang brauchen Sie vier:

| Typ | Bedeutung | Beispiel |
|-----|-----------|----------|
| `int` | ganze Zahl | `5`, `-273` |
| `float` | Kommazahl (Gleitkommazahl) | `3.14`, `2.5e5` |
| `bool` | Wahrheitswert | `True`, `False` |
| `str` | Zeichenkette (Text) | `"Wasser"` |

Den Typ eines Werts fragen Sie mit `type()` ab.

In [7]:
print(type(5))
print(type(5.0))
print(type(True))
print(type("Wasser"))

<class 'int'>
<class 'float'>
<class 'bool'>
<class 'str'>


Der Dezimaltrenner ist der **Punkt**, nicht das Komma. `3,14` ist für Python kein Kommawert,
sondern ein Paar aus zwei Zahlen — eine häufige Fehlerquelle beim Übertragen aus Excel.

Typen lassen sich umwandeln:

In [8]:
messwert_text = "78.4"           # z. B. aus einer Textdatei eingelesen
messwert = float(messwert_text)  # Umwandlung in eine Zahl

print(messwert + 1)              # jetzt kann man rechnen
print(int(7.9))                  # int() schneidet ab, es rundet NICHT

79.4
7


### Wahrheitswerte

Vergleiche liefern `True` oder `False`. Diese Werte werden in Teil 2 die Grundlage für Verzweigungen sein.

| Operator | Bedeutung |
|----------|-----------|
| `==` | gleich (zwei Gleichheitszeichen!) |
| `!=` | ungleich |
| `<`  `>` | kleiner, größer |
| `<=`  `>=` | kleiner gleich, größer gleich |

In [9]:
Re = 4200        # Reynolds-Zahl

print(Re > 2300)      # turbulent?
print(Re == 2300)
print(Re != 2300)

True
False
True


> **Klassischer Anfängerfehler:** `=` ist die Zuweisung, `==` ist der Vergleich. `Re = 2300` setzt
> die Reynolds-Zahl auf 2300, `Re == 2300` fragt, ob sie 2300 ist.

---
## 5. Ergebnisse sauber ausgeben: f-Strings

`print(n)` liefert `1.3739...` — 16 Stellen, keine Einheit, kein Kontext. Für ein Protokoll ist das
unbrauchbar. Mit einem **f-String** bauen Sie Werte direkt in Text ein: ein `f` vor dem
Anführungszeichen, die Variable in geschweiften Klammern.

In [10]:
print(f"Die Stoffmenge beträgt {n} mol.")

Die Stoffmenge beträgt 3.436544211141276 mol.


Die Anzahl der Nachkommastellen steuern Sie mit einem Formatzusatz nach einem Doppelpunkt:

| Format | Bedeutung | Ausgabe für `n = 1.37387...` |
|--------|-----------|------------------------------|
| `{n:.2f}` | 2 Nachkommastellen | `1.37` |
| `{n:.4f}` | 4 Nachkommastellen | `1.3739` |
| `{n:.3e}` | wissenschaftlich, 3 Stellen | `1.374e+00` |
| `{n:8.3f}` | 3 Nachkommastellen, 8 Zeichen breit | `   1.374` |

In [11]:
print(f"Stoffmenge:  n = {n:.3f} mol")
print(f"Druck:       p = {p:.3e} Pa")
print(f"Temperatur:  T = {T:.1f} K  =  {T - 273.15:.1f} °C")

Stoffmenge:  n = 3.437 mol
Druck:       p = 2.500e+05 Pa
Temperatur:  T = 350.0 K  =  76.9 °C


In den geschweiften Klammern darf auch gerechnet werden — im Beispiel oben die Umrechnung von
Kelvin in Grad Celsius. Halten Sie die Ausdrücke dort aber kurz; komplizierte Rechnungen gehören
in eine eigene Zeile davor.

### Beispiel: molare Masse einer Verbindung

Ein vollständiges kleines Programm — Eingabedaten, Rechnung, formatierte Ausgabe:

In [12]:
# molare Massen der Elemente in g/mol
M_C = 12.011
M_H = 1.008
M_O = 15.999

# Ethanol: C2H6O
M_ethanol = 2 * M_C + 6 * M_H + 1 * M_O

print(f"Molare Masse von Ethanol: M = {M_ethanol:.3f} g/mol")

# Wie viel Stoffmenge steckt in 250 g Ethanol?
m = 250.0                    # g
n_ethanol = m / M_ethanol    # mol

print(f"In {m:.0f} g Ethanol stecken n = {n_ethanol:.2f} mol.")

Molare Masse von Ethanol: M = 46.069 g/mol
In 250 g Ethanol stecken n = 5.43 mol.


---
## 6. Listen

Selten haben Sie nur einen Messwert. Eine **Liste** fasst mehrere Werte unter einem Namen zusammen.
Sie schreiben sie in eckigen Klammern, die Werte durch Kommata getrennt.

In [13]:
# Temperaturen an fünf Messstellen einer Kolonne, in °C
temperaturen = [78.4, 82.1, 91.5, 97.3, 99.8]

print(temperaturen)
print(f"Anzahl der Messstellen: {len(temperaturen)}")

[78.4, 82.1, 91.5, 97.3, 99.8]
Anzahl der Messstellen: 5


### Zugriff über den Index

Auf einzelne Elemente greifen Sie mit einem **Index** in eckigen Klammern zu. Entscheidend:
**Python zählt ab 0.** Das erste Element hat den Index 0, das fünfte den Index 4.

```
      temperaturen = [ 78.4 , 82.1 , 91.5 , 97.3 , 99.8 ]
Index                    0      1      2      3      4
von hinten              -5     -4     -3     -2     -1
```

In [14]:
print(temperaturen[0])    # erstes Element
print(temperaturen[2])    # drittes Element
print(temperaturen[-1])   # letztes Element
print(temperaturen[-2])   # vorletztes Element

78.4
91.5
99.8
97.3


### Bereiche herausschneiden (Slicing)

Mit `[start:stop]` schneiden Sie einen Teilbereich heraus. Der Startindex gehört dazu,
der Stoppindex **nicht** — `[1:3]` liefert die Elemente 1 und 2.

In [15]:
print(temperaturen[1:3])    # Elemente 1 und 2
print(temperaturen[:2])     # von Anfang bis ausschließlich 2
print(temperaturen[2:])     # ab Element 2 bis zum Ende

[82.1, 91.5]
[78.4, 82.1]
[91.5, 97.3, 99.8]


### Listen verändern

Listen sind veränderlich: Sie können Elemente überschreiben, anhängen und entfernen.

In [16]:
temperaturen[0] = 78.9          # Messwert korrigieren
temperaturen.append(101.2)      # neue Messstelle anhängen

print(temperaturen)
print(f"Maximum: {max(temperaturen)} °C, Minimum: {min(temperaturen)} °C")
print(f"Mittelwert: {sum(temperaturen) / len(temperaturen):.2f} °C")

[78.9, 82.1, 91.5, 97.3, 99.8, 101.2]
Maximum: 101.2 °C, Minimum: 78.9 °C
Mittelwert: 91.80 °C


> **Hinweis:** `sum`, `min`, `max` und `len` funktionieren direkt auf Listen. Für echtes Rechnen mit
> vielen Zahlen — elementweise multiplizieren, Kurven auswerten — sind Listen allerdings unbequem.
> Dafür lernen Sie in Teil 3 **NumPy-Arrays** kennen.

---
## 7. Dictionaries

Eine Liste greift über eine Position zu. Oft wollen Sie aber über einen **Namen** zugreifen:
"Wie groß ist die molare Masse von Wasser?" Dafür gibt es das **Dictionary** — eine Zuordnung von
**Schlüsseln** zu **Werten**, in geschweiften Klammern.

In [17]:
# Stoff -> molare Masse in g/mol
molare_massen = {
    "Wasser":   18.015,
    "Ethanol":  46.069,
    "Benzol":   78.114,
    "Toluol":   92.141,
}

print(molare_massen["Ethanol"])

46.069


In [18]:
# neuen Eintrag ergänzen
molare_massen["Aceton"] = 58.080

print(molare_massen)
print(f"Anzahl hinterlegter Stoffe: {len(molare_massen)}")
print("Wasser" in molare_massen)     # Prüfen, ob ein Schlüssel existiert

{'Wasser': 18.015, 'Ethanol': 46.069, 'Benzol': 78.114, 'Toluol': 92.141, 'Aceton': 58.08}
Anzahl hinterlegter Stoffe: 5
True


Dictionaries sind das natürliche Werkzeug für Stoffdaten. Sie können als Wert auch wieder ein
Dictionary hinterlegen und so kleine Stoffdatenbanken aufbauen:

In [19]:
stoffdaten = {
    "Wasser":  {"M": 18.015, "T_sied": 373.15, "rho": 998.0},
    "Ethanol": {"M": 46.069, "T_sied": 351.45, "rho": 789.0},
}

# Siedetemperatur von Ethanol
T_sied_ethanol = stoffdaten["Ethanol"]["T_sied"]

print(f"Ethanol siedet bei {T_sied_ethanol:.2f} K = {T_sied_ethanol - 273.15:.2f} °C")

Ethanol siedet bei 351.45 K = 78.30 °C


---
## 8. Übungen

Bearbeiten Sie die Aufgaben in der jeweils leeren Code-Zelle. Die Lösung können Sie darunter
aufklappen — versuchen Sie es aber zuerst selbst. Beim Programmieren lernt man ausschließlich
durch Tippen, nicht durch Lesen.

### Übung 1.1 — Dichte eines idealen Gases

Berechnen Sie die Dichte von Stickstoff (M = 28,014 g/mol) bei p = 3 bar und T = 400 K nach

$$\rho = \frac{p \, M}{R \, T}$$

Achten Sie auf die Einheiten: Rechnen Sie in Pa, kg/mol und K, damit die Dichte in kg/m³ herauskommt.
Geben Sie das Ergebnis mit zwei Nachkommastellen und Einheit aus.

In [20]:
# Ihre Lösung:

<details>
<summary><b>Lösung anzeigen</b></summary>

```python
p = 3e5                # Pa
T = 400.0              # K
M = 28.014e-3          # kg/mol  (aus g/mol durch Faktor 1e-3)
R = 8.314              # J/(mol K)

rho = p * M / (R * T)  # kg/m^3

print(f"Dichte von Stickstoff: rho = {rho:.2f} kg/m^3")
```

</details>

### Übung 1.2 — Reynolds-Zahl

In einem Rohr mit dem Innendurchmesser d = 25 mm strömt Wasser mit der Geschwindigkeit
w = 1,2 m/s. Dichte ρ = 998 kg/m³, dynamische Viskosität η = 1,002·10⁻³ Pa·s.

Berechnen Sie die Reynolds-Zahl

$$Re = \frac{\rho \, w \, d}{\eta}$$

und geben Sie zusätzlich mit einem Vergleichsoperator aus, ob sie über der kritischen Reynolds-Zahl
von 2300 liegt.

In [21]:
# Ihre Lösung:

<details>
<summary><b>Lösung anzeigen</b></summary>

```python
rho = 998.0       # kg/m^3
w   = 1.2         # m/s
d   = 25e-3       # m
eta = 1.002e-3    # Pa s

Re = rho * w * d / eta

print(f"Reynolds-Zahl: Re = {Re:.0f}")
print(f"Über der kritischen Reynolds-Zahl? {Re > 2300}")
```

</details>

### Übung 1.3 — Messreihe auswerten

An einem Rührkessel wurden folgende Temperaturen gemessen (in °C):

`[21.4, 24.8, 31.2, 38.7, 45.1, 49.9, 52.3]`

1. Speichern Sie die Werte in einer Liste.
2. Geben Sie den ersten und den letzten Messwert aus.
3. Berechnen Sie die Temperaturdifferenz zwischen letztem und erstem Wert.
4. Berechnen Sie den Mittelwert.
5. Geben Sie nur die ersten drei Messwerte aus.

In [22]:
# Ihre Lösung:

<details>
<summary><b>Lösung anzeigen</b></summary>

```python
T = [21.4, 24.8, 31.2, 38.7, 45.1, 49.9, 52.3]   # °C

print(f"Erster Wert:  {T[0]:.1f} °C")
print(f"Letzter Wert: {T[-1]:.1f} °C")

delta_T = T[-1] - T[0]
print(f"Aufheizspanne: {delta_T:.1f} K")

mittelwert = sum(T) / len(T)
print(f"Mittelwert: {mittelwert:.2f} °C")

print("Erste drei Messwerte:", T[:3])
```

</details>

### Übung 1.4 — Kleine Stoffdatenbank

Legen Sie ein Dictionary mit den molaren Massen von Methan (16,043 g/mol), Ethan (30,070 g/mol)
und Propan (44,097 g/mol) an.

1. Ergänzen Sie n-Butan mit 58,124 g/mol.
2. Berechnen Sie die Stoffmenge in 5 kg Propan.
3. Geben Sie das Ergebnis formatiert aus.

In [23]:
# Ihre Lösung:

<details>
<summary><b>Lösung anzeigen</b></summary>

```python
M = {
    "Methan": 16.043,
    "Ethan":  30.070,
    "Propan": 44.097,
}

M["n-Butan"] = 58.124

m_propan = 5000.0                    # g
n_propan = m_propan / M["Propan"]    # mol

print(f"In {m_propan/1000:.0f} kg Propan stecken n = {n_propan:.1f} mol.")
print("Hinterlegte Stoffe:", list(M.keys()))
```

</details>

---
## Zusammenfassung

| Konzept | Kurzform |
|---------|----------|
| Zuweisung | `T = 350.0   # K` |
| Potenz | `2 ** 10` |
| Ganzzahldivision / Rest | `7 // 2`, `7 % 2` |
| Modul importieren | `import math` |
| Typ abfragen | `type(x)` |
| Vergleich | `Re > 2300`, `x == y` |
| Formatierte Ausgabe | `f"T = {T:.2f} K"` |
| Liste | `werte = [1.0, 2.0, 3.0]`, `werte[0]`, `werte[1:3]` |
| Dictionary | `M = {"Wasser": 18.015}`, `M["Wasser"]` |

**Die drei häufigsten Anfängerfehler aus diesem Teil:**

1. Komma statt Punkt als Dezimaltrenner
2. `=` statt `==` beim Vergleichen
3. Index ab 1 statt ab 0 gezählt

---

**Weiter mit Teil 2:** Kontrollstrukturen, Funktionen und Fehlersuche.